src/ngx_http_lua_configureby.c

```diff
@@ -0,0 +1,287 @@
+
+/*
+ * Copyright (C) Yichun Zhang (agentzh)
+ *
+ * Author: Thibault Charbonnier (thibaultcha)
+ * I hereby assign copyright in this code to the lua-nginx-module project,
+ * to be licensed under the same terms as the rest of the code.
+ */
+
+
+#ifndef DDEBUG
+#define DDEBUG 0
+#endif
+
+
+#include "ddebug.h"
+#include "ngx_http_lua_configureby.h"
+#include "ngx_http_lua_directive.h"
+#include "ngx_http_lua_util.h"
+#include "ngx_http_lua_shdict.h"
+#include "ngx_http_lua_api.h"
+
+
+static ngx_conf_t *cfp;
+
+
+char *
+ngx_http_lua_configure_by_lua_block(ngx_conf_t *cf, ngx_command_t *cmd,
```

Comment:
I don't like all the code duplications in this C source file. It makes the code base harder to maintain and sync new changes.

Reply:
I can definitely improve it :) Which area? The context runner and/or the shm configure API?

Reply:
@thibaultcha Almost all the code in this C source file is duplicated C code.

Reply:
@agentzh I just added a new commit which introduces `ngx_http_lua_shared_dict_add`, which factorizes all of the shared code for shm create for `lua_shared_dict` and `ngx_configure.shared_dict()`.

As stated before, the code related to parsing and running the `_by_*_block` and `_by_*_file` directives of this module are already always duplicated - this feels outside the scope of this PR, but if you want me to address this, please advise.

Reply:
@thibaultcha 

>All of the Lua phases code is duplicated already, I just followed the pattern.

What exactly are you referring to?

The thing is if the config phase handlers can call these functions, then we should make it that way. I don't want to maintain 2 duplicated copies of configurations. Also, how do you handle config directive overriding and inheritance here? Seems like you already bypass the nginx config system entirely?

Reply:
> how do you handle config directive overriding and inheritance here? Seems like you already bypass the nginx config system entirely?

There is no inheritance involved here since we only support configuration of `http` (or above) level directives this is deliberate.

Ultimately, I do wish to support lower levels of configuration as explained in the PR (`server`, `location`...), but without it there are no need for supporting inheritance yet.

> What exactly are you referring to?

I was reffering to the only part of the code that I see is duplicated from other parts of ngx_lua, and that is the `ngx_http_lua_configure_by_lua` and other handlers that run the Lua code for this phase.

If not, then I am not sure what you are referring to when you are saying things are duplicated?